Author		: Joerg Friedrich
Date		: Thu, 15 Feb 2007 22:08:55 +0100
Message-ID	: <20070215210855.GA22572@stardust.friedrich-kn.de>
Status		: acked by David Miller, pushed upstream, confirmed working
Description	: fixes kenvctrld so it does not consume 100% CPU

diff -aur a/drivers/sbus/char/bbc_i2c.c b/drivers/sbus/char/bbc_i2c.c
--- a/drivers/sbus/char/bbc_i2c.c	2006-09-20 06:42:06.000000000 +0300
+++ b/drivers/sbus/char/bbc_i2c.c	2007-03-08 08:14:46.000000000 +0200
@@ -187,19 +187,18 @@
 	bp->waiting = 1;
 	add_wait_queue(&bp->wq, &wait);
 	while (limit-- > 0) {
-		u8 val;
+		long val;
 
-		set_current_state(TASK_INTERRUPTIBLE);
-		*status = val = readb(bp->i2c_control_regs + 0);
-		if ((val & I2C_PCF_PIN) == 0) {
+		val = wait_event_interruptible_timeout(bp->wq,
+			(((*status = readb(bp->i2c_control_regs + 0)) & I2C_PCF_PIN) == 0),
+			msecs_to_jiffies(250));
+		if ((val != -ERESTARTSYS) && (val > 0))  {
 			ret = 0;
 			break;
 		}
-		msleep_interruptible(250);
 	}
 	remove_wait_queue(&bp->wq, &wait);
 	bp->waiting = 0;
-	current->state = TASK_RUNNING;
 
 	return ret;
 }
@@ -340,7 +339,7 @@
 	 */
 	if (bp->waiting &&
 	    !(readb(bp->i2c_control_regs + 0x0) & I2C_PCF_PIN))
-		wake_up(&bp->wq);
+		wake_up_interruptible(&bp->wq);
 
 	return IRQ_HANDLED;
 }
